packages/contract/src/pay/DaimoPay.sol

```diff
@@ -0,0 +1,269 @@
+            intent: intent
+        });
+    }
+
+    // Claims a bridge transfer to its current recipient. If FastFinish happened
+    // for this transfer, then the recipient is the LP who fronted the amount.
+    // Otherwise, the recipient remains the original toAddr. The bridge transfer
+    // must already have been completed; coins are already in intent contract.
+    function claimAction(
+        PayIntent calldata intent,
+        Call[] calldata calls
+    ) public {
+        require(intent.toChainId == block.chainid, "DP: wrong chain");
+
+        PayIntentContract intentContract = intentFactory.createIntent(intent);
+
+        // Transfer from intent contract to this contract
+        intentContract.receiveAndSelfDestruct(intent);
+
+        // Finally, forward the balance to the current recipient
+        address recipient = intentToRecipient[address(intentContract)];
+        if (recipient == address(0)) {
+            // No LP showed up, so just complete the action.
+            recipient = intent.finalCall.to;
+
+            intentToRecipient[address(intentContract)] = recipient;
+            completeAction(address(intentContract), intent, calls);
+        } else {
+            // Otherwise, the LP fastFinished the action, give them the recieved
+            // amount.
+            intent.bridgeTokenOut.addr.safeTransfer(
+                recipient,
+                intent.bridgeTokenOut.amount
+            );
+        }
+
+        emit Claim({
+            intentAddr: address(intentContract),
+            finalRecipient: recipient,
+            intent: intent
+        });
+    }
+
+    // Swap bridgeTokenOut to finalCallToken
+    // Then, if the action has a finalCall, make the action calls.
+    // Otherwise, transfer the token to the action address.
+    function completeAction(
+        address intentAddr,
+        PayIntent calldata intent,
+        Call[] calldata calls
+    ) internal {
+        // Run arbitrary calls provided by the LP. These will generally approve
+        // the swap contract and swap if necessary
+        for (uint256 i = 0; i < calls.length; ++i) {
+            Call calldata call = calls[i];
+            (bool success, ) = call.to.call{value: call.value}(call.data);
+            require(success, "DP: swap call failed");
+        }
+
+        // Check that swap had a fair price
+        uint256 finalCallTokenBalance = TokenUtils.getBalanceOf(
+            intent.finalCallToken.addr,
+            address(this)
+        );
+        require(
+            finalCallTokenBalance >= intent.finalCallToken.amount,
+            "DP: insufficient final call token received"
+        );
+
+        if (intent.finalCall.data.length > 0) {
+            // If the intent is a call, approve the final token and make the call
+            TokenUtils.approve(
+                intent.finalCallToken.addr,
+                address(intent.finalCall.to),
+                intent.finalCallToken.amount
```

Comment:
in case finalCallTokenBalance is a bit higher than `intent.finalCallToken.amount`, what happens to the extra?

Reply:
one answer is that the LP just keeps it -> seems fine

Reply:
= we could do
```
finalCallToken.transfer(msg.sender, finalCallTokenBalance - intent.finalCallToken.amount)`

Reply:
in the success case